flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/StateLocalitySlotAssigner.java

```diff
@@ -139,6 +154,36 @@ public Collection<SlotAssignment> assignSlots(
         return assignments;
     }
 
+    /**
+     * The sorting principle and strategy here are very similar to {@link
```

Comment:
I'm not sure whether we want to prioritize local state over slot alignment only if the number of utilized task slots are equal. This slot assignment strategy is used if we want to minimize moving around state. What if we have a TM that has less slots but a bigger amount of state? In that case, the slot count would have higher priority than the state size. 🤔 

I think we might want to skip task slot prioritization in this implementation. WDYT?

Reply:
Let's discuss this topic in this thread.



Besides this comment https://github.com/apache/flink/pull/25218#issuecomment-2367911956 , I wanna add a case.

IIUC, this bug will be critical when the `taskmanager.numberOfTaskSlots` is high. For examples: `taskmanager.numberOfTaskSlots` is 10, and the job parallelism is 100, job needs 10 TMs.

- TM0 : runs subtask0 to subtask9
- TM1 : runs subtask10 to subtask19
- TM2 : runs subtask20 to subtask29
- ...
- TM9 : runs subtask90 to subtask99

When the parallelism is changed from 100 to 10. Ideally: job only needs 1 TMs.

- New subtask0 : processing data from old subtask0 to subtask9
- New subtask1 : processing data from old subtask10 to subtask19
- New subtask2 : processing data from old subtask20 to subtask29
- ...
- New subtask9 : processing data from old subtask90 to subtask99

But after state locality is applied: 

- The state of new subtask0 is from TM0
- The state of new subtask1 is from TM1
- The state of new subtask2 is from TM1
- ...
- The state of new subtask9 is from TM9

So, after rescaling, all TMs cannot be released. (It means the rescale is meaningless)

Reply:
> The user explicitly set local state recovery in the scenario where `StateLocalitySlotAssigner` is used (see [execution.state-recovery.from-local](https://github.com/apache/flink/blob/7adeecd3445947f42d3e3d1e2961b9464e910236/flink-core/src/main/java/org/apache/flink/configuration/StateRecoveryOptions.java#L108)), i.e. the user might value keeping the state on the machine in that case. 🤔

Yes, you are right. Another scenario is that users want to obtain the benefits of local recovery (quickly recover the state) without wasting resources.

If both of cases are needed, it seems we need an additional options to control it. (When using StateLocalitySlotAssigner, is local recovery high priority? Or is resources high priority?)

Anyway, after our discussion, it's better to only update the DefaultAssigneer in this PR. The strategy of `StateLocalitySlotAssigner` can be discussed in a separate JIRA or mail list.

WDYT?

> > The state locality only take effect during the job recovery, it's an optimization.
> 
> Why would that only have an affect during job recovery (i.e. when the Dispatcher recovers the job)? Every rescale operation recovers from a checkpoint in the end. Or am I misunderstanding you here?

Sorry, I didn't express clearly. I mean `job recovery` happens when job start or rescale.

Reply:
Thank you very much for the discussion !
It could be tracked in  https://issues.apache.org/jira/browse/FLINK-36426